pkg/controller/job/success_policy.go

```diff
@@ -0,0 +1,93 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package job
+
+import (
+	"fmt"
+
+	batch "k8s.io/api/batch/v1"
+	"k8s.io/apiserver/pkg/util/feature"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/features"
+)
+
+func canApplySuccessPolicy(jobCtx *syncJobCtx) bool {
+	return isIndexedJob(jobCtx.job) && feature.DefaultFeatureGate.Enabled(features.JobSuccessPolicy) &&
+		jobCtx.failureCondition == nil && jobCtx.successCondition == nil && jobCtx.job.Spec.SuccessPolicy != nil
+}
+
+func metSuccessPolicy(logger klog.Logger, successPolicy *batch.SuccessPolicy, completions int32, succeededIndexes orderedIntervals) (string, bool) {
+	if successPolicy == nil || len(succeededIndexes) == 0 {
+		return "", false
+	}
+
+	criteriaMetMsg := "Met criteria at index"
+	for index, criteria := range successPolicy.Criteria {
+		if criteria.SucceededIndexes != nil {
+			requiredIndexes := parseIndexesFromString(logger, *criteria.SucceededIndexes, int(completions))
+			// Failed to parse succeededIndexes of the criteria due to some errors like invalid format.
+			if len(requiredIndexes) == 0 {
+				continue
+			}
+			if metSucceededIndexesCriteria(requiredIndexes, succeededIndexes, criteria.SucceededCount) {
+				return fmt.Sprintf("%s %d", criteriaMetMsg, index), true
+			}
+		} else if criteria.SucceededCount != nil && succeededIndexes.total() >= int(*criteria.SucceededCount) {
+			return fmt.Sprintf("%s %d", criteriaMetMsg, index), true
+		}
+	}
+	return "", false
+}
+
+func metSucceededIndexesCriteria(requiredIndexes, succeededIndexes orderedIntervals, succeededCount *int32) bool {
+	if len(requiredIndexes) == 0 || len(succeededIndexes) == 0 {
+		return false
+	}
+	var contains, succeededPointer int
+	for _, criteria := range requiredIndexes {
+		for succeededPointer < len(succeededIndexes) {
+			// The range of criteria exists inside the range of succeededIndexes.
+			// 	(succeededIndexes[pointer].First <= criteria.First <= succeededIndexes[pointer].Last) AND
+			//	(succeededIndexes[pointer].First <= criteria.Last <= succeededIndexes[pointer].First)
+			if (criteria.First >= succeededIndexes[succeededPointer].First && criteria.First <= succeededIndexes[succeededPointer].Last) &&
```

Comment:
Isn't this equivalent to checking just `criteria.First >= succeededIndexes[succeededPointer].First && criteria.Last <= succeededIndexes[succeededPointer].Last`?

Reply:
Good point.
That's right. It indicates that the criteria exists the range of succeededIndexes[succeededPointer].

Reply:
Done.